lib/reversetunnel/conn.go

```diff
@@ -251,3 +254,77 @@ func (c *remoteConn) sendDiscoveryRequest(req discoveryRequest) error {
 
 	return nil
 }
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+const emitConnTargetPrefix = "Teleport"
+
+// emitConn is a wrapper for a net.Conn that emits an audit event for non-Teleport connections.
+type emitConn struct {
+	net.Conn
+	mu       *sync.RWMutex
+	buffer   bytes.Buffer
+	emitter  apievents.Emitter
+	ctx      context.Context
+	serverID string
+	emitted  bool
+}
+
+func newEmitConn(ctx context.Context, conn net.Conn, emitter apievents.Emitter, serverID string) *emitConn {
+	return &emitConn{
+		Conn:     conn,
+		mu:       &sync.RWMutex{},
+		buffer:   bytes.Buffer{},
+		emitter:  emitter,
+		ctx:      ctx,
+		serverID: serverID,
+		emitted:  false,
+	}
+}
+
+func (conn *emitConn) Read(p []byte) (int, error) {
+	conn.mu.RLock()
+	n, err := conn.Conn.Read(p)
+
+	// skip buffering if already could have emitted or will never emit
+	if err != nil || conn.buffer.Len() == len(emitConnTargetPrefix) || conn.serverID == "" {
+		conn.mu.RUnlock()
+		return n, err
+	}
+	conn.mu.RUnlock()
+
+	conn.mu.Lock()
+	defer conn.mu.Unlock()
+
+	remaining := len(emitConnTargetPrefix) - conn.buffer.Len()
+	_, err = conn.buffer.Write(p[:min(n, remaining)])
+	if err != nil {
+		return n, err
+	}
+
+	if conn.buffer.Len() == len(emitConnTargetPrefix) && !bytes.HasPrefix(conn.buffer.Bytes(), []byte(emitConnTargetPrefix)) {
```

Comment:
Let me see if I get this right: we emit the event unless we the first thing we read from the connection is the string "Teleport"?

Could you explain here if that is the case and why that works.